### PR TITLE
Update Dependency Syncing Algorithm and Infrastructure

### PR DESCRIPTION
**Pull request by NickGerleman**
Facebook has been changing the structure of the OSS react-native repo, and we have to make changes to accommodate these. The major changes are around syncing dependencies for monorepo packages, along with Facebook's introduction of the `repo-config` package that hoists devDependencies we want to sync.

This change revamps the dependency syncing process to accommodate these changes and improve other nits we've discovered.

- We now mirror the repo-config package in our repo so that we can examine new RN dev dependencies
- We properly sync dependencies for non-react-native monorepo packages with override manifests
- The dependency syncing algorithm is now unit testable and has tests
- We are more liberal sycning dev dependencies, synchronizing all matching packages in our repo to the version in core if it is newer than what we have
- We don't make breaking changes to peer dependencies when integrating builds into stable branches

Some related changes were made to override tooling to fix some pain points when bringing in monorepo packages, including repo-config. Namely:
- We move from override validation using Lerna to the tool itself being able to reason about monorepos. This means we no longer have to fork the package.json of RN packages to add override validation scripts.
- Multi-manifest support leads to some mechanical changes to the API layer to make it a bit cleaner
- Override upgrades can also work on the entire monorepo
- Output is now clearer when running validation on monorepo packages
- We exempt node_modules directories from inclusion in override manifests by default

Manually validated integrating a new build using this logic and added UTs

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6254)

### CODE REVIEW

File: packages/@react-native/repo-config/package.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  "dependencies": {
+    "@babel/core": "^7.0.0",
+    "@babel/generator": "^7.5.0",
+    "@reactions/component": "^2.0.2",
+    "async": "^2.4.0",
+    "babel-eslint": "^10.1.0",
+    "clang-format": "^1.2.4",
+    "connect": "^3.6.5",
+    "coveralls": "^3.0.2",
+    "detox": "15.4.4",
```
Comment (NickGerleman) [inline packages/@react-native/repo-config/package.json:20]: I was a bit concerned adding this as a package into our repo would bloat things, but the extra tree of dependencies was less bad than I expected.

File: packages/@rnw-scripts/find-repo-root/src/findRepoRoot.ts [hunk 2]
```diff
@@ -7,24 +7,15 @@
...
-      const packagePath = path.join(dir, 'package.json');
-      if (!(await findUp.exists(packagePath))) {
-        return undefined;
-      }
-
-      const pkg = (await fs.promises.readFile(packagePath)).toString();
-      return JSON.parse(pkg).name === 'react-native-windows-repo'
-        ? dir
-        : undefined;
+      return (await findUp.exists(path.join(dir, '.git'))) ? dir : undefined;
```
Comment (NickGerleman) [inline packages/@rnw-scripts/find-repo-root/src/findRepoRoot.ts:18]: Needed to change this for the monorepo support of react-native-platform-override

File: packages/@rnw-scripts/integrate-rn/src/integrateRN.ts [hunk 3]
```diff
@@ -59,16 +56,16 @@ import findRepoRoot from '@rnw-scripts/find-repo-root';
...
+async function enumerateOverridePackages(): Promise<WritableNpmPackage[]> {
+  return await enumerateLocalPackages(isOverridePackages);
 }
 
 /**
- * Does the NPM package correspond to an out-of-tree platform?
+ * Whether the NPM package is subject to override validation
  */
-async function isOutOfTreePlatform(pkg: NpmPackage): Promise<boolean> {
+async function isOverridePackages(pkg: NpmPackage): Promise<boolean> {
```
Comment (NickGerleman) [inline packages/@rnw-scripts/integrate-rn/src/integrateRN.ts:68]: nit: typo

File: package.json [hunk 4]
```diff
@@ -13,7 +13,7 @@
     "format": "format-files -i -style=file -assume-filename=../.clang-format",
     "format:verify": "format-files -i -style=file -assume-filename=../.clang-format -verify",
     "test": "lerna run test --parallel -- --color",
-    "validate-overrides": "lerna run validate-overrides --parallel -- --color"
+    "validate-overrides": "react-native-platform-override validate"
```
Comment (NickGerleman) [inline package.json:16]: Need to still pass the color flag

File: packages/react-native-platform-override/src/FileSearch.ts [hunk 5]
```diff
@@ -27,6 +28,29 @@ export async function findManifest(cwd?: string): Promise<string> {
...
+  if (!packagePath) {
+    throw new Error('This command must be run within a package');
+  }
+
+  const packageDir = path.dirname(packagePath);
+  if (await exists(path.join(packageDir, 'overrides.json'))) {
+    return [path.join(packageDir, 'overrides.json')];
+  }
+
+  const localPackages = await enumerateLocalPackages(pkg =>
```
Comment (NickGerleman) [inline packages/react-native-platform-override/src/FileSearch.ts:47]: This won't correctly use the passed in optional cwd right now

File: packages/@rnw-scripts/integrate-rn/src/integrateRN.ts [hunk 6]
```diff
@@ -15,19 +13,18 @@ import * as yargs from 'yargs';
...
-import {exec, ExecOptions} from 'child_process';
 import {
   upgradeOverrides,
   validateManifest,
   ValidationError,
 } from 'react-native-platform-override';
 
-import findRepoRoot from '@rnw-scripts/find-repo-root';
+import runCommand from './runCommand';
+import {upgradeDependencies} from './upgradeDependencies';
```
Comment (NickGerleman) [inline packages/@rnw-scripts/integrate-rn/src/integrateRN.ts:27]: Should change this to be a default export.

**Comment by TheSavior** [comment]
Having gone through this I'd love to hear your thoughts on the repo reorganization. Has it been helpful? Are there parts that are strictly worse? Other things that would make it even better?

**Comment by NickGerleman** [comment]
@TheSavior Overall I think package separation makes sense in a lot of cases, especially for bits like RNTester that conceptually consume react-native instead of being part of it.

From the purely OSS perspective, I don't like how the `repo-config` package relies on hoisting dependencies for other packages to pick up. That pattern could be a bit fragile or hard to reason about, but it sounded like there were some internal needs that required it and that it wasn't the end goal.

In general I think we pushed more for self-contained packages in RNW for better or worse. E.g. each package has its own devDependencies, eslint config, tsconfig, flowconfig, etc. Generally config files extend from shared packages to avoid duplication. This has the pros of allowing more scoped changes (e.g. increasing or lowering strictness per package instead of the repo), and makes things easier to reason about. The downside is that we have a lot of boilerplate, and a lot of dependencies to keep in sync between packages. We've definitely had a moment before of "Why do we have two copies of Babel in our lockfile?". Some of our internal repos have tooling to help with this.

I could see issues with the current structure around the intersection of stable branches with monorepo packages RN depends on. I.e. if a breaking change is made to a package in master, then we want to cherry-pick a fix to an older branch, the options for the current scheme seem non-ideal. For RNW scoped packages (currently just [`@react-native-windows/cli`](https://www.npmjs.com/package/@react-native-windows/cli)) we follow the same overall versioning scheme of the main package of distinct master vs stable branch versions to get around this. I also get the feeling manual version bumping/publish will become more of a pain in general as the number of packages increases.

File: packages/react-native-platform-override/src/e2etest/Api.test.ts [hunk 7]
```diff
@@ -32,67 +34,72 @@ test('validateManifest', async () => {
...
-      await Api.removeOverride('ReactCommon\\yoga\\yoga\\Yoga.cpp', opts),
+      await Api.removeOverride(
+        'ReactCommon\\yoga\\yoga\\Yoga.cpp',
+        manifestPath,
+      ),
     ).toBe(true);
 
     expect(
-      await Api.removeOverride('ReactCommon\\yoga\\yoga\\Karate.cpp', opts),
+      await Api.removeOverride(
```
Comment (NickGerleman) [inline packages/react-native-platform-override/src/e2etest/Api.test.ts:64]: I should change these to all take the manifest path as the first arg

**Review by rectified95 — APPROVED** [review]

**Comment by NickGerleman** [comment]
Oh wow I hit "push" instead of "push to" in VSCode and it dropped all of the commits from this directly into master.

Seems like we should maybe have a branch restriction to prevent that.

**Pull request merged**